be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/LaoValidator.scala

```diff
@@ -1,140 +1,328 @@
 import ch.epfl.pop.storage.DbActor
 
 import scala.concurrent._
+import scala.util.matching.Regex
 import scala.util.{Failure, Success}
 
-case object LaoValidator extends MessageDataContentValidator {
-  def validateCreateLao(rpcMessage: JsonRpcRequest): GraphMessage = {
-    def validationError(reason: String): PipelineError = super.validationError(reason, "CreateLao", rpcMessage.id)
-
-    rpcMessage.getParamsMessage match {
-      case Some(message: Message) =>
-        val data: CreateLao = message.decodedData.get.asInstanceOf[CreateLao]
-        val expectedHash: Hash = Hash.fromStrings(data.organizer.base64Data.toString, data.creation.toString, data.name)
-
-        val channel: Channel = rpcMessage.getParamsChannel
-
-        if (!validateTimestampStaleness(data.creation)) {
-          Right(validationError(s"stale 'creation' timestamp (${data.creation})"))
-        } else if (!validateWitnesses(data.witnesses)) {
-          Right(validationError("duplicate witnesses keys"))
-        } else if (expectedHash != data.id) {
-          Right(validationError("unexpected id"))
-        } else if (data.organizer != message.sender) {
-          Right(validationError("unexpected organizer public key"))
-        } else if (channel != Channel.ROOT_CHANNEL) {
-          Right(validationError(s"trying to send a CreateLao message on a channel $channel other than ${Channel.ROOT_CHANNEL}"))
-        } else if (data.name == "") {
-          Right(validationError(s"LAO name must not be empty"))
-        } else {
-          Left(rpcMessage)
-        }
-      case _ => Right(validationErrorNoMessage(rpcMessage.id))
+object LaoValidator extends MessageDataContentValidator {
+
+  val laoValidator = new LaoValidator(DbActor.getInstance)
+
+  def validateCreateLao(rpcMessage: JsonRpcRequest): GraphMessage = laoValidator.validateCreateLao(rpcMessage)
+
+  def validateStateLao(rpcMessage: JsonRpcRequest): GraphMessage = laoValidator.validateStateLao(rpcMessage)
+
+  def validateGreetLao(rpcMessage: JsonRpcRequest): GraphMessage = laoValidator.validateGreetLao(rpcMessage)
+
+  def validateUpdateLao(rpcMessage: JsonRpcRequest): GraphMessage = laoValidator.validateUpdateLao(rpcMessage)
+
+  private val dupWitnessError = "duplicate witnesses keys"
+  private val unexpectedID = "unexpected id"
+  private val addressRegPat = "^.*:\\/\\/[a-zA-Z0-9.]+:?[0-9]*[a-zA-Z0-9\\/]*".r
+
+  sealed class LaoValidator(dbActorRef: => AskableActorRef) extends MessageDataContentValidator {
+
+    def validateCreateLao(rpcMessage: JsonRpcRequest): GraphMessage = {
+      def validationError(reason: String): PipelineError = super.validationError(reason, "CreateLao", rpcMessage.id)
+
+      rpcMessage.getParamsMessage match {
+        case Some(message: Message) =>
+          val (createLao, _, senderPK, channel) = extractData[CreateLao](rpcMessage)
+          val expectedHash: Hash = Hash.fromStrings( // needs Checking in docs or protocol.
```

Comment:
Is this comment still relevant ? I am not sure if keeping it would be a good idea. It seems as it is something the user should check (I am thinking of future students who will see the code).
I guess you put it here when you refactored as a reminder. Correct me if I am wrong.

Reply:
It was for myself indeed, fixing it now !